### PR TITLE
Field names test

### PR DESCRIPTION
**Pull request by amareesh86**
This PR tests for the generation of field names by CsvScheme.

### CODE REVIEW

**Comment by amareesh86** [comment]
@kuujo can you review?

**Comment by kuujo** [comment]
The `getHeaderRecord` provides behavior that is redundant to the `detectHeaders` method. Both can/should be done at the same time and should only load the header once. Both checks can use the same call to `getHeaderRecord`

**Comment by kuujo** [comment]
@amareesh86 shouldn't this have failing tests now? Or have field/header intersection tests not been written yet?

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 1]
```diff
@@ -824,13 +827,28 @@ public String getNullString() {
...
+
+
+    if (!getSourceFields().isUnknown()) {
+      Fields sourceFields = getSourceFields();
+
+      if (validateFields(flowProcess, tap, sourceFields)) {
+        return sourceFields;
+      } else {
+        throw new RuntimeException("Fields count don't match column count in input file");
+      }
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: I don't believe this block is going to work as expected. `Fields.UNKNOWN` is tantamount to passing a `null` or empty `Fields` object. So, the `Fields` size will be `0` and we can only really gleam information from the headers if they're provided.

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 2]
```diff
@@ -824,13 +827,28 @@ public String getNullString() {
...
+
+    }
+
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: Nevermind :-P

**Comment by kuujo** [comment]
@amareesh86 I think there may be a couple more useful assertions for fields that can be added. Let's try to establish the rules for fields and columns:
* If `Fields` is `UNKNOWN` and headers is `null`/empty and `skipHeaderRecord` is false, assign positional names to the columns and perform no validation
* If `Fields` is `UNKNOWN` and headers is `null`/empt but `skipHeaderRecord` is true, detect the headers and perform no further validation
* If `Fields` are defined, headers is `null`/empty and `skipHeaderRecord` is true, detect the header and verify that the `Fields` names are present in the header
* If `Fields` are defined, headers is `null`/empty and `skipHeaderRecord` is false, verify that the number of fields matches the number of columns 
* If `Fields` are defined and headers are defined, verify that the headers match the number of fields in the first record, and verify that the field names are present in the header names. `skipHeaderRecord` does not contribute to this decision
* If `Fields` are `UNKNOWN` and headers are defined, verify that the headers match the number of fields in the first record and assign header names to `Fields`. Again, `skipHeaderRecord` is consequential only to parsing in this case since headers are specified.

Does that make sense? Did I miss any case?

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 3]
```diff
@@ -824,15 +533,37 @@ public String getNullString() {
...
+      }
+
+    } else if (fields.isUnknown() && format.getHeader() != null) {
+      if (validateHeaders(flowProcess, tap)) {
+
+        setSourceFields(new Fields(format.getHeader()));
+
+      } else {
+        throw new RuntimeException("Headers count don't match column count in input file");
+      }
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: This is a group of pretty complex logic. Can we get some inline documentation on the conditions used here?

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 4]
```diff
@@ -824,15 +533,37 @@ public String getNullString() {
...
-      return getSourceFields();
+    Fields fields = getSourceFields();
 
-    if (format.getSkipHeaderRecord() && format.getHeader() == null) {
+    if (fields.isUnknown() && format.getHeader() == null && format.getSkipHeaderRecord()) {
       setSourceFields(detectHeader(flowProcess, tap, false));
-    } else if (format.getHeader() != null) {
-      setSourceFields(new Fields(format.getHeader()));
-    } else {
+    } else if (fields.isUnknown() && format.getHeader() == null && !format.getSkipHeaderRecord()) {
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: According to IntelliJ, `format.getSkipHeaderRecord()` is always `true` here

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 5]
```diff
@@ -824,15 +533,37 @@ public String getNullString() {
...
+      if (fields.size() != format.getHeader().length) {
+        throw new RuntimeException("Fields count don't match header count");
+      }
+
+    } else if (!fields.isUnknown() && format.getHeader() != null) {
+      if (!(validateFields(flowProcess, tap, fields) && validateFieldsAndHeaders(fields, Arrays.asList(format.getHeader())))) {
+        throw new RuntimeException("Headers count don't match column count in input file");
+      }
+
+    } else if (fields.isUnknown() && format.getHeader() != null) {
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: According to IntelliJ, `format.getHeader()` is always not `null` here

File: src/main/java/com/datascience/cascading/scheme/CsvScheme.java [hunk 6]
```diff
@@ -842,25 +573,97 @@ public Fields retrieveSourceFields(FlowProcess<JobConf> flowProcess, Tap tap) {
...
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Validates that fields properly reference headers.
+   */
+  protected boolean validateFieldsAndHeaders(Fields fields, List<String> headersList) {
+    if (headersList.size() != fields.size()) {
```
Comment (kuujo) [inline src/main/java/com/datascience/cascading/scheme/CsvScheme.java]: This is validating that the number of user-provided fields matches the number of user-provided headers right? If so, that's not correct.

```
If Fields are defined and CSVFormat.getHeader() are defined, the scheme will verify that the provided headers match the number of values in the first record in the CSV file, and verify that the provided field names are all present in the header names.
```

What we need to verify is that the user provided headers match the number of columns in the file regardless of whether `skipHeaderRecord` is `true`. That may be done elsewhere, but my point is that `Fields` *can* be less than the number of headers, but not greater than. So, I think this should be `if (fields.size() > headersList.size()) return false`. The validation properly implemented in `validateHeaders` should really be the only identity validation where `headerRecord.size() == format.getHeader().length`

**Comment by kuujo** [comment]
This is good :+1:

**Pull request merged**